### PR TITLE
Task/WG-61-Vite environments for local development

### PR DESCRIPTION
**Pull request by iamthatian**
## Overview: ##
This is an attempt to add environments in the react code using the `vite` build system.

On build/run, `vite` will read the `.env` file

This behavior is defined in `vite.config.ts`.

#### .ENV
And the `GEOAPI_BACKEND`  defined in `.env` will determine the target backend for local development. So, one would now have to edit the `.env` file instead of the `environments.ts` to test.

I've also included the `JWT` inside the `.env` and there is an `.env.example` for reference (`.env` is included in `.gitignore`).

#### Environment Variables
The environment variables defined in `vite.config.ts` will be exposed in the react code directly via an ~__ENV__ const (e.g. `__ENV__.apiUrl`).~ Changed to `process.env` (it seems to be this is generally compatible with tests (like jest), otherwise, because of [vite's build process](https://jestjs.io/docs/29.1/getting-started#using-vite) the `defines` are not picked up by jest, so I've opted to use `process.env` convention)

This interface is set in `src/vite.env.d.ts` for intellisense.

#### Concerns
~One concern is that now the running environment is not dynamically determined by the hostname on runtime (e.g. `hazmapper.local` or `hazmapper.utexas.edu`) and is predetermined with the prefix. I haven't thought of any issues during the build process. Would there be some cases this could be problematic? 

Another concern with this is with local development. So, we might have to enforce `hazmapper.local` instead of `localhost` for the correct `clientId`.~
Only for local development for now.

## PR Status: ##

* [X] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

[WG-61](https://jira.tacc.utexas.edu/browse/WG-61)

## Testing Steps: ##
1. Copy the `.env.example` to `.env` in the `JWT`, `GEOAPI_BACKEND`, (to `development`) in `.env`.
2. Ensure that `npm run dev` works.
3. Go to http://hazmapper.local:4200.
4. Ensure that you see the local `apiUrl` display correctly next to **Menu** and that environment variables are accessible via `process.env`.

## UI Photos:

N/A

### CODE REVIEW

File: react/vite.config.ts [hunk 1]
```diff
@@ -1,11 +1,94 @@
-import { defineConfig } from 'vite';
+import { defineConfig, loadEnv } from 'vite';
 import react from '@vitejs/plugin-react';
 
+function getApiUrl(backend: string): string {
```
Comment (nathanfranklin) [inline react/vite.config.ts:4]: I think should be renamed `getGeoApiUrl` and add a comment

File: react/vite.config.ts [hunk 2]
```diff
@@ -1,11 +1,94 @@
...
+  } else if (backend === 'staging') {
+    return 'https://agave.designsafe-ci.org/geo-staging/v2';
+  } else if (backend === 'production') {
+    return 'https://agave.designsafe-ci.org/geo/v2';
+  } else {
+    throw new Error('Unsupported Type');
+  }
+}
+
+function getPortalUrl(backend: string): string {
```
Comment (nathanfranklin) [inline react/vite.config.ts:16]: rename to something more explicit like getDesignSafePortalUrl?

maybe add comment?

File: react/src/vite-env.d.ts [hunk 3]
```diff
@@ -1 +1,26 @@
 /// <reference types="vite/client" />
+
+interface Env {
+  designSafeUrl: string;
+  backend: string;
+  apiUrl: string;
+  portalUrl: string;
```
Comment (nathanfranklin) [inline react/src/vite-env.d.ts:7]: i think these should be more explicit as both are used as apis and we got so many portals out there. how about:

apiUrl -> geoApiUrl
portalUrl -> designSafePortalUrl

File: react/vite.config.ts [hunk 4]
```diff
@@ -1,11 +1,94 @@
...
+    env.mapillaryClientId = '4936281379826603';
+    env.mapillaryClientToken =
+      'MLY|4936281379826603|f8c4732d3c9d96582b86158feb1c1a7a';
+  } else {
+    env.apiUrl = getApiUrl(envFile.BACKEND);
+    env.portalUrl = getPortalUrl(envFile.BACKEND);
+    env.clientId = 'Eb9NCCtWkZ83c01UbIAITFvhD9ka';
+    env.host = 'hazmapper.local';
+    env.baseHref = '/';
+    env.mapillaryClientSecret =
```
Comment (nathanfranklin) [inline react/vite.config.ts:75]: can you check these mapillary values?  looks the same as prod?

File: react/src/vite-env.d.ts [hunk 5]
```diff
@@ -1 +1,26 @@
...
+  clientId: string;
+  host: string;
+  baseHref: string;
+  jwt: string;
```
Comment (nathanfranklin) [inline react/src/vite-env.d.ts:11]: jwt will need to also handle null: `string | null;`

File: README.md [hunk 6]
```diff
@@ -15,6 +15,35 @@ See https://github.com/TACC-Cloud/geoapi which is an associated restful API.
...
+
+#### Environments
+
+Environments are handled by vite via `vite.config.ts` and the `.env` files.
+
+
+The vite commands defined in `package.json` with the `--mode` prefix determines which environment we are running/building in.
+
+
+Then, for local development, the `BACKEND` specified in `.env` will allow testing different backends.
```
Comment (nathanfranklin) [inline README.md:27]: should we split this BACKEND variable into two different ones:
* GEOAPI_BACKEND
* DESIGNSAFE_PORTAL

maybe for the short term this suggestion isn't so important. thoughts?  i can just make a follow on ticket.

**Review by nathanfranklin — CHANGES_REQUESTED** [review]
I like the idea 👍   gonna discuss the staging/prod approach in slack.

I left some initial comments but getting this error when running `npm run dev ` or `npm run build:local`

```failed to load config from /Users/nathanf/dev/hazmapper/react/vite.config.ts
error during build:
Error: Unsupported Type
    at getApiUrl (file:///Users/nathanf/dev/hazmapper/react/vite.config.ts.timestamp-1681358144738.mjs:12:11)
    at file:///Users/nathanf/dev/hazmapper/react/vite.config.ts.timestamp-1681358144738.mjs:61:18
    at loadConfigFromFile (file:///Users/nathanf/dev/hazmapper/react/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:63023:15)
    at async resolveConfig (file:///Users/nathanf/dev/hazmapper/react/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:62643:28)
    at async doBuild (file:///Users/nathanf/dev/hazmapper/react/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:45554:20)
    at async build (file:///Users/nathanf/dev/hazmapper/react/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:45543:16)
    at async CAC.<anonymous> (file:///Users/nathanf/dev/hazmapper/react/node_modules/vite/dist/node/cli.js:756:9)```

File: README.md [hunk 7]
```diff
@@ -15,6 +15,35 @@ See https://github.com/TACC-Cloud/geoapi which is an associated restful API.
...
+
+
+First, to set the target backend for local development, edit the `.env` file and add the target backend.
+Possible target backends are `development`, `staging`, and `production`.
+```
+BACKEND="TARGET BACKEND"
+```
+
+
+Then add the jwt retreived from [Getting started](###getting-started) to `.env`.
```
Comment (nathanfranklin) [inline README.md:37]:
```suggestion
Then add the jwt retrieved from [Getting started](###getting-started) to `.env`.
```

File: README.md [hunk 8]
```diff
@@ -15,6 +15,35 @@ See https://github.com/TACC-Cloud/geoapi which is an associated restful API.
 
 `react/` has the react client
 
+
+#### Environments
```
Comment (nathanfranklin) [inline README.md:19]: good docs here 💯  👍

**Review by nathanfranklin — CHANGES_REQUESTED** [review]
Some additional comments to consider

I'm reaching out on slack about the use of .env in our two environments using the same image.

**Review by sophia-massie — APPROVED** [review]
LGTM! Thank you!!

**Pull request merged**